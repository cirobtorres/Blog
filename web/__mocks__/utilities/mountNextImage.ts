const nextOptimizedImage = (filePath: string) => {
  if (filePath) {
    const filename = filePath.substring(
      filePath.lastIndexOf("/") + 1,
      filePath.lastIndexOf(".")
    );
    const extension = filePath.substring(filePath.lastIndexOf(".") + 1);
    const path = filePath.substring(0, filePath.lastIndexOf("/") + 1);
    const URL = `${path}${filename}.${extension}`;
    const encodedURL = encodeURIComponent(URL);
    const width = "w=3840";
    const quality = "q=75";
    const nextOptimizedImage = `/_next/image?url=${encodedURL}&${width}&${quality}`;
    return nextOptimizedImage;
  }
  return null;
};

interface ToEncodeURIAttributes {
  imageFilename: string;
  imageFileExtension: string;
  path: string;
  width: number;
}

function mountLocalImage({
  imageFilename,
  imageFileExtension,
  path,
  width,
}: ToEncodeURIAttributes) {
  // Mock local images
  const quality = 75;
  const URL = `${path}${imageFilename}.${imageFileExtension}`;
  const encodedURL = encodeURIComponent(URL);
  const nextOptimizedImage = `/_next/image?url=${encodedURL}&w=${width}&q=${quality}`;
  return nextOptimizedImage;
}

function requestBackEndImage(
  filePath?: string,
  width: number = 1920,
  options?: {
    PROTOCOL?: string;
    IP?: string;
    PORT?: string;
  }
) {
  // Mock images managed by Strapi
  const quality = 75;

  if (filePath) {
    const filename = filePath.substring(
      filePath.lastIndexOf("/") + 1,
      filePath.lastIndexOf(".")
    );
    const extension = filePath.substring(filePath.lastIndexOf(".") + 1);
    const path = filePath.substring(0, filePath.lastIndexOf("/") + 1);

    const protocol = options ? options.PROTOCOL : "http";
    const ip = options ? options.IP : "127.0.0.1";
    const port = options ? options.PORT : "1337";
    const backendURL = `${protocol}://${ip}:${port}`;

    const URL = `${backendURL}${path}${filename}.${extension}`;

    const encodedURL = encodeURIComponent(URL);
    const nextOptimizedImage = `/_next/image?url=${encodedURL}&w=${
      width * 2
    }&q=${quality}`;

    return nextOptimizedImage;
  }

  const URL = "/images/not-authenticated.png";
  const encodedURL = encodeURIComponent(URL);

  const nextOptimizedImage = `/_next/image?url=${encodedURL}&w=${
    width * 2
  }&q=${quality}`;

  return nextOptimizedImage;
}

export { nextOptimizedImage, mountLocalImage, requestBackEndImage };
