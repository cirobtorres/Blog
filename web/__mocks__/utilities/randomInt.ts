import { faker } from "@faker-js/faker";

export const randomInt = (min: number, max: number) =>
  faker.number.int({ min, max });
