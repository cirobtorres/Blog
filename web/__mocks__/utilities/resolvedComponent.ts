async function resolvedComponent(Component: {
  (): Promise<React.JSX.Element>;
}) {
  const ComponentResolved = await Component();
  return () => ComponentResolved;
}

async function resolvedComponentWithProps<T extends object>(
  Component: (props: T) => React.JSX.Element | Promise<React.JSX.Element>,
  props: T
) {
  const ComponentResolved = await Component(props);
  return () => ComponentResolved;
}

export { resolvedComponent, resolvedComponentWithProps };
