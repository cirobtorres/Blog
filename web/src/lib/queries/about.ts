const GET_ABOUT = `
query About {
  about {
    documentId
    title
    display_name
    github_link
    github_blog_link
    media {
      documentId
      url
      alternativeText
    }
    updatedAt
    createdAt
    publishedAt
    blocks {
      ... on ComponentSharedRichText {
        id
        body
      }
      ... on ComponentSharedMedia {
        id
        file {
          documentId
          url
          alternativeText
          caption
          width
          height
        }
      }
      ... on ComponentSharedSlider {
        id
        files {
          documentId
          url
          alternativeText
          caption
          width
          height
        }
      }
    }
  }
}`;

export { GET_ABOUT };
