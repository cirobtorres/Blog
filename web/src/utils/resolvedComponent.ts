/* eslint-disable @typescript-eslint/no-explicit-any */
async function resolvedComponent(Component: {
  (): Promise<React.JSX.Element>;
}) {
  const ComponentResolved = await Component();
  return () => ComponentResolved;
}

async function resolvedComponentWithProps(
  Component: ({ props }: { [key: string]: any }) => React.JSX.Element,
  props: { [key: string]: any }
) {
  const ComponentResolved = Component({ ...props });
  return () => ComponentResolved;
}

export default resolvedComponent;
export { resolvedComponentWithProps };
