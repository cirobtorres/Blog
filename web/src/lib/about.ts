import graphqlClient from "./graphQlClient";
import { GET_ABOUT } from "./queries/about";

const getAbout = async () => {
  try {
    const { about }: { about: About } = await graphqlClient.request(GET_ABOUT);
    return { data: about };
  } catch (error) {
    console.error("Failed to fetch about:", error);
    throw new Error("Failed to fetch about");
  }
};

export { getAbout };
