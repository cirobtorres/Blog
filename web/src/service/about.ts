import graphqlClient from "../lib/graphQlClient";
import { GET_ABOUT } from "../lib/queries/about";

const getAbout = async () => {
  try {
    const { about }: { about: About } = await graphqlClient.request(GET_ABOUT);
    return { data: about };
  } catch (error) {
    console.error("Failed to fetch about:", error);
    throw new Error("Failed to fetch about");
  }
};

export { getAbout };
