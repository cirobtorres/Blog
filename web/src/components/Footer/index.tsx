import { getAbout } from "@/service/about";
import Link from "next/link";

const Footer = async () => {
  const {
    data: { github_blog_link: link },
  } = await getAbout();
  return (
    <footer
      data-testid="footer"
      aria-label="Footer com link do código fonte"
      className="shrink-0 h-60 border-y border-blog-border bg-blog-background-2"
    >
      <div className="h-full flex flex-col justify-center items-center gap-4 mx-auto max-w-screen-2xl">
        <div className="flex justify-center">
          <p
            data-testid="footer-paragraph"
            className="max-w-[75%] text-center text-wrap text-sm"
          >
            Encontrou algum mau funcionamento neste site?
            <br />
            <Link
              data-testid="footer-link"
              href={link || "/"}
              target="_blank"
              rel="noopener noreferrer"
              aria-label="Repositório do Github para o código fonte do blog"
              className="break-words transition-colors duration-500 text-blog-foreground-highlight hover:text-blog-foreground-readable-hover"
            >
              {link || "/"}
            </Link>
          </p>
        </div>
        <div className="w-full max-w-44 h-1 inline-grid">
          <div className="h-full col-start-1 row-start-1 bg-gradient-to-r from-transparent to-blog-foreground-highlight blur-lg rounded-full" />
          <div className="h-full col-start-1 row-start-1 bg-gradient-to-r from-transparent to-blog-foreground-highlight rounded-full" />
        </div>
      </div>
      <div className="py-1">
        <p className="text-center text-xs text-[#5a5a5a]">
          Powered by Strapi & Next.js
        </p>
      </div>
    </footer>
  );
};

export default Footer;
