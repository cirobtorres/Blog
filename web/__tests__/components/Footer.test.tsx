import { axe } from "jest-axe";
import { toHaveNoViolations } from "jest-axe";
import { act, render, screen, waitFor } from "@testing-library/react";
import Footer from "../../src/components/Footer";
import React from "react";
import { resolvedComponent } from "../../__mocks__/utilities/resolvedComponent";
import { getAbout } from "@/service/about";

jest.mock("../../src/service/about", () => ({
  getAbout: jest.fn(),
}));

function returnMock(props?: {
  documentId?: string;
  title?: string;
  github_link?: string;
  github_blog_link?: string;
  createdAt?: string;
  updatedAt?: string;
  publishedAt?: string;
  blocks?: ArticleBlocks[];
}) {
  return (getAbout as jest.Mock).mockResolvedValue({
    data: {
      documentId: "12345",
      title: "My Blog",
      github_link: "https://github.com/johndoe",
      github_blog_link: "https://github.com/johndoe/blog",
      createdAt: "2025-03-10T23:11:17.381Z",
      updatedAt: "2025-03-10T23:11:17.381Z",
      publishedAt: "2025-03-10T23:11:17.381Z",
      blocks: [],
      ...props,
    },
  });
}

describe("Footer", () => {
  it("renders the footer component", async () => {
    returnMock();
    const FooterResolved = await resolvedComponent(Footer);
    render(<FooterResolved />);
    await waitFor(() => {
      const footer = screen.getByTestId("footer");
      expect(footer).toBeInTheDocument();
    });
  });

  it("renders the footer attributes", async () => {
    returnMock();
    const FooterResolved = await resolvedComponent(Footer);
    render(<FooterResolved />);
    await waitFor(() => {
      const footer = screen.getByTestId("footer");
      expect(footer).toHaveAttribute(
        "aria-label",
        "Footer com link do código fonte"
      );
    });
  });

  it("renders the correct github footer link", async () => {
    returnMock();
    const FooterResolved = await resolvedComponent(Footer);
    render(<FooterResolved />);
    await waitFor(() => {
      const link = screen.getByTestId("footer-paragraph")?.querySelector("a");
      expect(link).toHaveTextContent("https://github.com/johndoe/blog");
      expect(link).toHaveAttribute("href", "https://github.com/johndoe/blog");
      expect(link).toHaveAttribute("target", "_blank");
      expect(link).toHaveAttribute("rel", "noopener noreferrer");
      expect(link).toHaveAttribute(
        "aria-label",
        "Repositório do Github para o código fonte do blog"
      );
      expect(link).toHaveAttribute("href", "https://github.com/johndoe/blog");
    });
  });

  it("should be accessible", async () => {
    returnMock();
    const FooterResolved = await resolvedComponent(Footer);
    const { container } = render(<FooterResolved />);
    // Either act or waitFor are necessary here because of changing of states
    await act(async () => {
      const results = await axe(container);
      expect.extend(toHaveNoViolations);
      expect(results).toHaveNoViolations();
    });
  });

  it("is responsive", async () => {
    returnMock();
    const FooterResolved = await resolvedComponent(Footer);
    render(<FooterResolved />);
    await waitFor(() => {
      const footerParagraph = screen.getByTestId("footer-paragraph");
      expect(footerParagraph).toHaveClass("text-wrap");
      const link = screen.getByTestId("footer-link");
      expect(link).toHaveClass(
        "break-words text-blog-foreground-highlight hover:text-blog-foreground-readable-hover"
      );
    });
  });

  it("links to home page if footer link is not set on the server", async () => {
    returnMock({
      github_link: "",
      github_blog_link: "",
    });
    const FooterResolved = await resolvedComponent(Footer);
    render(<FooterResolved />);
    await waitFor(() => {
      const link = screen.getByTestId("footer-link");
      expect(link.getAttribute("href")).toBe("/");
    });
  });

  it("matches the snapshot", async () => {
    returnMock();
    const FooterResolved = await resolvedComponent(Footer);
    const { asFragment } = render(<FooterResolved />);
    await waitFor(() => expect(asFragment()).toMatchSnapshot());
  });
});
