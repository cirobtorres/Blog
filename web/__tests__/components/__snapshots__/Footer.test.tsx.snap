// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Footer matches the snapshot 1`] = `
<DocumentFragment>
  <footer
    aria-label="Footer com link do código fonte"
    class="shrink-0 h-60 border-y border-blog-border bg-blog-background-2"
    data-testid="footer"
  >
    <div
      class="h-full flex flex-col justify-center items-center gap-4 mx-auto max-w-screen-2xl"
    >
      <div
        class="flex justify-center"
      >
        <p
          class="max-w-[75%] text-center text-wrap text-sm"
          data-testid="footer-paragraph"
        >
          Encontrou algum mau funcionamento neste site?
          <br />
          <a
            aria-label="Repositório do Github para o código fonte do blog"
            class="break-words transition-colors duration-500 text-blog-foreground-highlight hover:text-blog-foreground-readable-hover"
            data-testid="footer-link"
            href="https://github.com/johndoe/blog"
            rel="noopener noreferrer"
            target="_blank"
          >
            https://github.com/johndoe/blog
          </a>
        </p>
      </div>
      <div
        class="w-full max-w-44 h-1 inline-grid"
      >
        <div
          class="h-full col-start-1 row-start-1 bg-gradient-to-r from-transparent to-blog-foreground-highlight blur-lg rounded-full"
        />
        <div
          class="h-full col-start-1 row-start-1 bg-gradient-to-r from-transparent to-blog-foreground-highlight rounded-full"
        />
      </div>
    </div>
    <div
      class="py-1"
    >
      <p
        class="text-center text-xs text-[#5a5a5a]"
      >
        Powered by Strapi & Next.js
      </p>
    </div>
  </footer>
</DocumentFragment>
`;
